File: python/fitness_function.py
"""fitness_function.py
Evaluate a parameter set against ground-truth cyst counts.

This is a direct Python translation of fitness_function.m.
"""

from __future__ import annotations
import numpy as np
from concurrent.futures import ProcessPoolExecutor
from detect_cyst_peaks import detect_cyst_peaks


def _worker(args):
    mask_path, params = args
    return detect_cyst_peaks(mask_path, params)


def evaluate_params(
    params: list | np.ndarray,
    gt_masks: list[str],
    gt_counts: list[int] | np.ndarray,
    n_workers: int = -1,
) -> tuple[float, dict]:
    """Evaluate a parameter set against all training masks.

    Parameters
    ----------
    params    : 6-element sequence – see detect_cyst_peaks for definitions.
    gt_masks  : list of paths to NIfTI mask files.
    gt_counts : ground-truth cyst counts (same order as gt_masks).
    n_workers : number of parallel workers (-1 = all CPUs).

    Returns
    -------
    score   : MAPE (%) – value to minimise.
    metrics : dict with per-case and aggregate statistics.
    """
    gt = np.asarray(gt_counts, dtype=float)
    n  = len(gt_masks)

    import os
    if n_workers == -1:
        n_workers = os.cpu_count() or 1

    args = [(m, params) for m in gt_masks]
    if n_workers == 1:
        # Sequential: avoids spawning subprocesses that each load large NIfTI files
        predictions = np.array([_worker(a) for a in args], dtype=float)
    else:
        with ProcessPoolExecutor(max_workers=n_workers) as pool:
            predictions = np.array(list(pool.map(_worker, args)), dtype=float)

    errors       = np.abs(predictions - gt)
    pct_error    = (predictions - gt) / gt
    abs_pct_err  = np.abs(pct_error)
    overest      = np.maximum(0, predictions - gt)
    underest     = np.maximum(0, gt - predictions)

    MAPE      = abs_pct_err.mean() * 100
    MPE       = pct_error.mean()   * 100
    bias_pen  = max(0, abs(MPE) - 10) * 2
    MAE       = errors.mean()
    STD       = errors.std(ddof=1)
    RMSE      = np.sqrt((errors ** 2).mean())
    MAD       = np.median(errors)
    SS_res    = ((gt - predictions) ** 2).sum()
    SS_tot    = ((gt - gt.mean())   ** 2).sum()
    R2        = 1 - SS_res / SS_tot if SS_tot > 0 else float('nan')
    mean_bias = (predictions - gt).mean()
    ratio_over = overest.sum() / (underest.sum() + 1e-6)

    score = MAPE   # Option A (MAPE only) – matches MATLAB default

    metrics = {
        'MAPE':         MAPE,
        'MPE':          MPE,
        'MAE':          MAE,
        'STD':          STD,
        'RMSE':         RMSE,
        'MAD':          MAD,
        'R_squared':    R2,
        'mean_bias':    mean_bias,
        'ratio_over':   ratio_over,
        'predictions':  predictions,
        'ground_truth': gt,
        'errors':       errors,
        'score_components': {'primary': MAPE, 'bias_penalty': bias_pen},
    }
    return score, metrics

File: matlab/fitness_function.m
function [score, metrics] = fitness_function(params, gt_masks, gt_am)
% fitness_function  Evaluate a parameter set against ground-truth cyst counts.
%
%   [score, metrics] = fitness_function(params, gt_masks, gt_am)
%
%   params   : 1×6 vector – see detectCystPeaks for parameter definitions.
%   gt_masks : N×1 cell array of full paths to NIfTI mask files.
%   gt_am    : 1×N vector of ground-truth cyst counts.
%
%   score    : Mean Absolute Percentage Error (MAPE, %) – the quantity
%              minimised by the genetic algorithm.
%   metrics  : Struct with per-sample predictions and aggregate statistics.

    numCases       = numel(gt_am);
    predictions    = zeros(numCases, 1);
    errors         = zeros(numCases, 1);
    overestimation = zeros(numCases, 1);
    underestimation= zeros(numCases, 1);
    percent_error  = zeros(numCases, 1);
    abs_pct_error  = zeros(numCases, 1);

    parfor n = 1:numCases
        pred             = detectCystPeaks(gt_masks{n}, params);
        predictions(n)   = pred;
        errors(n)        = abs(pred - gt_am(n));
        overestimation(n)  = max(0,  pred - gt_am(n));
        underestimation(n) = max(0,  gt_am(n) - pred);
        percent_error(n)   = (pred - gt_am(n)) / gt_am(n);
        abs_pct_error(n)   = abs(percent_error(n));
    end

    % ---- Metrics ----
    MAPE      = mean(abs_pct_error) * 100;
    MPE       = mean(percent_error) * 100;
    bias_penalty = max(0, abs(MPE) - 10) * 2;
    MAE       = mean(errors);
    STD_err   = std(errors);
    RMSE      = sqrt(mean(errors .^ 2));
    MAD       = median(errors);

    SS_res    = sum((gt_am(:) - predictions(:)) .^ 2);
    SS_tot    = sum((gt_am(:) - mean(gt_am(:)))   .^ 2);
    R_squared = 1 - SS_res / SS_tot;

    mean_bias = mean(predictions(:) - gt_am(:));
    ratio_over = sum(overestimation) / (sum(underestimation) + 1e-6);

    % ---- Fitness (Option A – MAPE only, recommended for publication) ----
    score = MAPE;

    % Option B – MAPE + bias penalty for |MPE| > 10 %:
    % score = MAPE + bias_penalty;

    % ---- Output struct ----
    metrics = struct();
    metrics.MAPE       = MAPE;
    metrics.MPE        = MPE;
    metrics.MAE        = MAE;
    metrics.STD        = STD_err;
    metrics.RMSE       = RMSE;
    metrics.MAD        = MAD;
    metrics.R_squared  = R_squared;
    metrics.mean_bias  = mean_bias;
    metrics.ratio_over = ratio_over;
    metrics.predictions  = predictions(:)';
    metrics.ground_truth = gt_am(:)';
    metrics.errors       = errors(:)';
    metrics.score_components = struct('primary', MAPE, 'bias_penalty', bias_penalty);
end
